SC-31975: Elastic cookbook for RAG pipeline (#172)
rungalileo/docs-official: cookbooks/use-cases/rag-elastic-langchain-integration.mdx

---
title: "Elasticsearch Integration for Retrieval-Augmented Generation (RAG)"

> description: "Guide to using Elasticsearch with LangGraph for the Chatbot RAG app, logging to Galileo."

Reviewer: ```suggestion
description: "Guide to using Elasticsearch with LangGraph for the Chatbot RAG app, logging to Galileo."
tag: "Python"
```

We started tagging the cookbooks by language, so adding a tag here

---

> ## Elasticsearch + RAG + Galileo Overview

Reviewer: ```suggestion
```
We don't need a title here - this will come from the page title.

Elasticsearch, with capabilities like the ELSER model for sparse vector retrieval, allows you to build powerful semantic search and RAG applications. This guide demonstrates how to integrate your Elasticsearch-backed knowledge base with a LangGraph agent and stream observability data to **Galileo**.

In this guide you'll:

1.  Set up Elasticsearch as a vector store for your RAG application.
2.  Configure Galileo for logging.
3.  Stream LangGraph traces to **Galileo** for end-to-end observability.

## Set up Elasticsearch

> Ensure your Elasticsearch instance is running and accessible. Fetch your API keys based on [this guid]e(https://www.elastic.co/getting-started/enterprise-search/vector-search)

Reviewer: ```suggestion
Ensure your Elasticsearch instance is running and accessible. Fetch your API keys based on [this guide](https://www.elastic.co/getting-started/enterprise-search/vector-search)
```

 You'll need to configure the `ElasticsearchStore` from `langchain-elasticsearch` to connect to your index.

> You can find the implementation of an example [chat app](https://github.com/rungalileo/elasticsearch-labs/blob/main/example-apps/chatbot-rag-app/api/chat.py).

Reviewer: Can we move this to the SDK examples repo: https://github.com/rungalileo/sdk-examples. We should have all SDK example code in the same repo

# store_retriever = vector_store.as_retriever()
```

This setup allows your LangGraph agent to retrieve relevant documents from Elasticsearch using the ELSER model.

## Use Galileo for logging

To monitor your RAG application with Galileo, you'll need to set up the Galileo callback handler and include it in your LangGraph invocations.

First, ensure your Galileo environment variables are set. These are typically found in a `.env` file or your environment settings (as seen in `env.example`):

```python
import os
# GalileoCallback is usually provided by the galileo_observe package
from galileo.handlers.langchain import GalileoCallback # Or your specific import path for GalileoCallback

# --- Galileo Environment Variables ---
# These should be set in your environment.
# Example values from the Elastic Chatbot RAG app's env.example:
os.environ["GALILEO_API_KEY"] = os.getenv("GALILEO_API_KEY", "<your-galileo-api-key>")
os.environ["GALILEO_PROJECT_NAME"] = os.getenv("GALILEO_PROJECT_NAME", "elastic-chat-rag-app")
# os.environ["GALILEO_CONSOLE_URL"] = os.getenv("GALILEO_CONSOLE_URL", "https://console.galileo.ai/") # Often optional

> os.environ["GALILEO_LOG_STREAM"] = "elastic_rag_stream_demo" # Define a stream name for this application

Reviewer: Why is this setting environment variables to environment variables? This is also encouraging setting API Keys in code which is bad practice.

There's no loading of the environment variables, so don't you need to use dotenv here?

Author: added dotenv

# Initialize the Galileo handler
# In the Elastic Chatbot RAG app, this might be done in an 'observe.py' module
# and imported as 'galileo_handler'
galileo_handler = GalileoCallback()
```

Once the `galileo_handler` is initialized (or imported, as in `api/chat.py` which uses `from observe import galileo_handler`), you can add it to your LangGraph agent's configuration when invoking it:

> ```python

Reviewer: I can't find this code in the linked sample

Author: removed

### Imports and Configuration

First, we import the necessary libraries and configure our environment variables. This part of the script loads your API keys and sets up the connection details for Elasticsearch, OpenAI, and Galileo.

<SnippetElasticRag1 />

### 1. Elasticsearch Setup
<SnippetElasticRag2 />
The code automatically:
- Connects to your Elasticsearch instance
- Connects to the ELSER model for semantic search
- Creates an index and stores sample documents

### 2. Agent Architecture
<SnippetElasticRag3 />
- **State Management**: Uses `AgentState` to track conversation messages
- **Tool Integration**: Creates a retriever tool that searches Elasticsearch
- **LangGraph Workflow**: Defines the flow between agent reasoning and tool usage

### 3. Conversation Flow
<SnippetElasticRag4 />
1. User asks a question
2. Agent decides whether to use the retriever tool
3. If needed, searches Elasticsearch for relevant documents
4. Generates a response based on retrieved context
5. Saves the conversation to chat history

### 4. Adding Galileo observability

1. Open your [Galileo Console](https://app.galileo.ai/)
2. Navigate to your project (f.e. `elasticsearch-rag-demo`)

> 3. You'll see traces for each question, showing:

Reviewer: At this point I'm not seeing traces as there is nothing running. This should come after we run the project
